compare-flatcar-with-gentoo

```diff
@@ -0,0 +1,742 @@
+    local ver="${1}"
+    local arr_name="${2}"
+
+    # strip everything after first underline or dash, whichever comes
+    # first, so for strings like 1.2.3_alpha or 1.2.3-r2 we get 1.2.3
+    ver="${ver%%[-_]*}"
+    IFS='.' read -ra "${arr_name}" <<< "${ver}"
+
+    local i
+    local -n arr_ref="${arr_name}"
+
+    for i in $(seq 0 "$((${#arr_ref[@]}-1))"); do
+        strip_leading_zeros "${arr_name}[${i}]"
+    done
+}
+
+# last version field may have a letter
+split_last_version_field() {
+    local field="${1}"
+    local num_name="${2}"
+    local word_name="${3}"
+
+    local -n num_ref="${num_name}"
+    local -n word_ref="${word_name}"
+
+    num_ref="${field%%*([a-z])}"
+    word_ref="${field##*([0-9])}"
+
+    strip_leading_zeros "${num_name}"
+}
+
+get_suffixes() {
+    local ver="${1}"
+    local arr_name="${2}"
+    local ver_tmp
+
+    if [[ ! "${ver}" =~ _ ]]; then
+        local -n arr_ref="${arr_name}"
+        arr_ref=()
+        return
+    fi
+    ver_tmp="${ver#*_}"
+    ver="${ver_tmp%-*}"
+    IFS='_' read -ra "${arr_name}" <<< "${ver}"
+}
+
+split_suffix() {
+    local suffix="${1}"
+    local word_name="${2}"
+    local num_name="${3}"
+
+    local -n word_ref="${word_name}"
+    local -n num_ref="${num_name}"
+
+    word_ref="${suffix%%*([0-9])}"
+    num_ref="${suffix##*([a-z])}"
+
+    strip_leading_zeros "${num_name}"
+}
+
+get_revision() {
+    local version="${1}"
+    local rev_name="${2}"
+
+    local -n rev_ref="${rev_name}"
+
+    if [[ ! "${version}" =~ '-' ]]; then
+        rev_ref='0'
+        return
+    fi
+    rev_ref="${version##*-r}"
+
+    strip_leading_zeros "${rev_name}"
+}
+
+declare -A SUFFIXES=(
+    ['alpha']='0'
+    ['beta']='1'
+    ['pre']=2
+    ['rc']=3
+    # empty string can't be a key
+    # ['']=4
+    ['p']=5
+)
+
+get_suffix_word_index() {
+    local word="${1}"
+    local index_ref="${2}"
+
+    local -n index="${index_ref}"
+
+    # we can't put an empty key into the hash map
+    if [[ -z "${word}" ]]; then
+        index=4
+        return
+    fi
+    # assign -1 to invalid suffix
+    index="${SUFFIXES[${word}]--1}"
+}
+
+gentoo_vercmp() {
+    local a_ver="${1}"
+    local b_ver="${2}"
+    local res_name="${3}"
+
+    local -n res_ref="${res_name}"
+
+    local -a a_nums
+    local -a b_nums
+
+    get_plain_version "${a_ver}" a_nums
+    get_plain_version "${b_ver}" b_nums
+
+    if [[ "${#a_nums[@]}" -ne "${#b_nums[@]}" ]]; then
+        if [[ "${#a_nums[@]}" -gt "${#b_nums[@]}" ]]; then
+            append_with "${#a_nums[@]}" 0 b_nums
+        else
+            append_with "${#b_nums[@]}" 0 a_nums
+        fi
+    fi
+
+    local i
+    local a
+    local b
+
+    # last field needs to be checked separately
+    for i in $(seq 0 $(("${#a_nums[@]}" - 2))); do
+        a="${a_nums[${i}]}"
+        b="${b_nums[${i}]}"
+
+        if [[ "${a}" -gt "${b}" ]]; then
+            res_ref=">v${i}n"
+            return
+        fi
+        if [[ "${a}" -lt "${b}" ]]; then
```

Comment:
here too `line 278: [[: 18b: value too great for base (error token is "18b")`

Reply:
Fixed too.